README.md
# unlock_puppet task

#### Table of Contents

1. [Description](#description)
1. [Setup](#setup)
1. [Usage - Basic](#usage)
1. [Reference - Parameters](#reference)
1. [Alternate Usage](#alternate-usage)
1. [Getting Help - With Tasks](#getting-help)

## Description

This module provides an `unlock_puppet` class and task that ...

* Kills `puppet agent` runs exceeding the maximum of `runinterval` or `runtimeout`.
* Restarts the `Puppet Agent` service (if it is enabled) if the last run report exceeds the maximum of `runinterval` or `runtimeout`.
* Restarts the `Puppet Agent` and/or `PXP Agent` service if it is not running.

This is valuable when a puppet agent process is locked, and/or the puppet or pxp-agent service needs to be restarted.

## Setup

* Install the module

## Usage

### Automated Enforcement

* Apply the `unlock_puppet` class to a node

```puppet
include unlock_puppet
```

The `unlock_puppet` class will create a cron job (or scheduled task) to resolve a locked `puppet agent` process, and/or a stopped `Puppet Agent` or `PXP Agent` service.

### Ad-Hoc Enforcement

The `unlock_puppet` task resolve a locked `puppet agent` process, and/or a stopped `Puppet Agent` service, but will not resolve a stopped `PXP Agent` service.

* Use the `puppet task run` command

```bash
puppet task run unlock_puppet --nodes agent.example.com
```

```bash
[root@master]# puppet task run unlock_puppet --nodes agent.example.com
Starting job ...
Note: The task will run only on permitted nodes.
New job ID: 1
Nodes: 1

Started on agent.example.com ...
Finished on node agent.example.com
  result : checking puppet agent process and service, puppet agent process lock file age 86400 exceeds runinterval 1800 or runtimeout 3600, killing puppet agent process, deleting puppet agent process lock file
  status : success

Job completed. 1/1 nodes succeeded.
Duration: 1 sec
```

Rather than specifying the nodes directly via `--nodes` you could use `--query` to query for nodes than have not reported within a cutoff:

```bash
# calculate a `not-responding` datestamp
export CUTOFF_DATE=$(date -u +"%Y-%m-%dT%H:%M:%SZ" -d "-$(puppet config print runinterval) seconds")

# validate the datestamp using `puppet query`:
date
echo $CUTOFF_DATE
puppet query "nodes { report_timestamp < '$CUTOFF_DATE' }"

# run the task with the query
puppet task run unlock_puppet --query "nodes { report_timestamp < '$CUTOFF_DATE' }"
```

## Reference

### Class Parameters

Note that class parameter(s) are only necessary for resource-like class declarations, which are only necessary to uninstall the resources of this module.

#### ensure

Data type: String

This parameter is optional, with valid options of 'present' (the default) and 'absent'.

To uninstall the cron job (or scheduled task) apply the class using a resource-like class declaration:

```
class { 'unlock_puppet': ensure => absent }
```

### Task Parameters

#### force_agent

Boolean, default: false

Ignore `runinterval` and `runtimeout`, kill the puppet agent process, and delete its lock file.

#### force_service

Boolean, default: false

Ignore `runinterval` and `runtimeout`, and restart the puppet service (even if it is not enabled).

```bash
[root@master]# puppet task run unlock_puppet --nodes agent.example.com force_agent=true force_service=true
Starting job ...
Note: The task will run only on permitted nodes.
New job ID: 2
Nodes: 1

Started on agent.example.com ...
Finished on node agent.example.com
  result : checking puppet agent process and service, stopping puppet service, killing puppet agent process, deleting puppet agent process lock file, starting puppet service
  status : success

Job completed. 1/1 nodes succeeded.
Duration: 4 sec
```

## Alternate Usage

The script executed by this task can be extracted from the `files` directory of this module and run locally on the command line:

```bash
[root@agent]# /usr/local/bin/unlock_puppet.rb
{"status":"success","result":"checking puppet agent process and service, puppet agent process lock file age 86400 exceeds runinterval 1800 or runtimeout 3600, killing puppet agent process, deleting lock file"}
```

The script accepts the same parameters as the task via the following command line options:

```bash
[root@agent]# /usr/local/bin/unlock_puppet.rb --force_agent --force_service
{"status":"success","result":"checking puppet agent process and service, stopping puppet service, killing puppet agent process, deleting puppet agent process lock file, starting puppet service"}
```

## Getting Help

To show help for tasks, run `puppet task run --help`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## Release 2.0.8

- resolve lint and rubocop issues

## Release 2.0.7

- specify the full path to the puppet binary

## Release 2.0.6

- do not read STDIN in the script executed by cron

## Release 2.0.5

- manage the pxp-agent service, except when run as a task

## Release 2.0.4

- handle windows service start_pending and stop_pending states

## Release 2.0.3

- use maximum of runinterval or runtimeout

## Release 2.0.2

- add an ensure parameter to allow uninstalls of the module

## Release 2.0.1

- start service if not running

## Release 2.0.0

- refactor
- rename parameters
- document classes
- adjust schedules

## Release 1.0.4

- clarify documentation

## Release 1.0.3

- update windows support

## Release 1.0.2

- document query

## Release 1.0.1

- document reoccurring task

## Release 1.0.0

- correct windows support
- implement reoccurring task

## Release 0.1.0

**Features**

**Bugfixes**

**Known Issues**


tasks/init.rb
#!/opt/puppetlabs/puppet/bin/ruby

# unlock_puppet/tasks/init.rb and unlock_puppet/files/unlock_puppet.rb are similar,
# except that the task does not interact with the pxp-agent service, and raises errors,
# and the script does not read JSON from STDIN.

require 'json'
require 'facter'
require 'puppet'

####

def read_params
  options = {}
  begin
    Timeout.timeout(2) do
      options = JSON.parse(STDIN.read)
    end
  rescue Timeout::Error
    require 'optparse'
    options['force_process'] = 'false'
    options['force_service'] = 'false'
    parser = OptionParser.new do |opts|
      opts.banner = 'Usage: unlock_puppet.rb [options]'
      opts.separator ''
      opts.separator 'Summary: Unlock puppet agent runs exceeding runinterval or runtimeout'
      opts.separator ''
      opts.separator 'Options:'
      opts.separator ''
      opts.on('--force_process', 'Force a kill of the puppet agent process, and delete its lock file') do
        options['force_process'] = 'true'
      end
      opts.on('--force_service', 'Force a restart of the puppet service') do
        options['force_service'] = 'true'
      end
      opts.on('-h', '--help', 'Display help') do
        puts opts
        puts
        exit 0
      end
    end
    parser.parse!
  end
  options
end

params = read_params

force_process = params['force_process'] == 'true'
force_service = params['force_service'] == 'true'

Puppet.initialize_settings

lockfile      = Puppet[:agent_catalog_run_lockfile]
lastrunreport = Puppet[:lastrunreport]
runinterval   = Puppet[:runinterval] || 1800
runtimeout    = Puppet[:runtimeout]  || 1800

result = {}
report = []

####

def puppet_binary
  (Facter.value(:os)['family'] == 'windows') ? 'puppet' : '/opt/puppetlabs/bin/puppet'
end

def write_log(message)
  message.tr!('"', "'")
  if Facter.value(:os)['family'] == 'windows'
    `powershell.exe " If ([System.Diagnostics.EventLog]::SourceExists(\"Unlock Puppet\") -eq $False) { New-EventLog -Source \"Unlock Puppet\" -LogName Application } "`
    `powershell.exe " Write-EventLog -Source \"Unlock Puppet\" -LogName Application -EntryType Information -EventID 1234 -Message \"unlock_puppet: #{message}\" "`
  else
    `logger -p local3.info "#{message}"`
  end
end

def kill_process(pid)
  return if pid.zero?
  if Facter.value(:os)['family'] == 'windows'
    `taskkill /f /pid #{pid}`
  else
    `kill -9 #{pid}`
  end
end

def start_service(service)
  `#{puppet_binary} resource service #{service} ensure=running`
  $?
end

def stop_service(service)
  if Facter.value(:os)['family'] == 'windows'
    result = `SC QUERY #{service} | FIND "STATE"`
    if $?.exitstatus.zero?
      service_state = result.split(' ').last.chomp
      if service_state == 'START_PENDING' || service_state == 'STOP_PENDING'
        result = `SC QUERYEX #{service} | FIND "PID"`
        if $?.exitstatus.zero?
          service_pid = result.split(' ').last.chomp.to_i
          if service_pid > 0
            kill_process(service_pid)
            return $?
          end
        end
      end
    end
    `SC STOP #{service}`
  else
    `#{puppet_binary} resource service #{service} ensure=stopped`
  end
  $?
end

def service_status(service)
  `#{puppet_binary} resource service #{service}`
end

def service_enabled(status)
  status.downcase.include?('true')
end

def service_running(status)
  status.downcase.include?('running')
end

####

begin
  report << 'checking puppet agent process and service'

  if force_service
    report << 'stopping puppet service'
    stop_service('puppet')
  end

  if File.file?(lockfile)
    run_pid = File.read(lockfile).to_i
    if force_process
      report << 'killing puppet agent process'
      kill_process(run_pid)
      if File.file?(lockfile)
        report << 'deleting puppet agent process lock file'
        File.delete(lockfile) if File.file?(lockfile)
        raise StandardError('unable to delete puppet agent process lock file') if File.file?(lockfile)
      end
    else
      lockfile_age = (Time.now - File.stat(lockfile).mtime).to_i
      if lockfile_age > [runinterval, runtimeout].max
        report << "puppet agent process lock file age #{lockfile_age} exceeds runinterval #{runinterval} or runtimeout #{runtimeout}"
        report << 'killing puppet agent process'
        kill_process(run_pid)
        if File.file?(lockfile)
          report << 'deleting puppet agent process lock file'
          File.delete(lockfile) if File.file?(lockfile)
          raise StandardError('unable to delete puppet agent process lock file') if File.file?(lockfile)
        end
      end
    end
  end

  puppet_service_status = service_status('puppet')

  if force_service
    report << 'starting puppet service'
    command = start_service('puppet')
    raise StandardError('unable to start puppet service') unless command.exitstatus.zero?
  elsif service_enabled(puppet_service_status)
    if service_running(puppet_service_status)
      if File.file?(lastrunreport)
        lastrunreport_age = (Time.now - File.stat(lastrunreport).mtime).to_i
        # Either [runinterval, runtimeout].max or runinterval
        if lastrunreport_age > [runinterval, runtimeout].max
          report << "last run report age #{lastrunreport_age} exceeds runinterval #{runinterval} or runtimeout #{runtimeout}"
          report << 'stopping puppet service'
          stop_service('puppet')
          report << 'starting puppet service'
          command = start_service('puppet')
          raise StandardError('unable to start puppet service') unless command.exitstatus.zero?
        end
      else
        report << 'last run report absent'
        report << 'stopping puppet service'
        stop_service('puppet')
        report << 'starting puppet service'
        command = start_service('puppet')
        raise StandardError('unable to start puppet service') unless command.exitstatus.zero?
      end
    else
      # status = puppet_service_status.match(/ensure.*?,/m).to_s.split("'")[1]
      # report << "puppet service status: #{status}"
      report << 'puppet service not running'
      report << 'stopping puppet service'
      stop_service('puppet')
      report << 'starting puppet service'
      command = start_service('puppet')
      raise StandardError('unable to start puppet service') unless command.exitstatus.zero?
    end
  end

  # To enable notifications, create unlock_notify.rb in this directory, containing a unlock_notify(result) method.
  # The unlock_notify(result) method must accept one parameter: the 'result' hash.
  notify_library = File.join(__dir__, 'unlock_notify.rb')
  if File.file?(notify_library)
    require notify_library
  end

  result['status'] = 'success'
  unlock_notify(result) if defined?(unlock_notify)
  result['result'] = report.join(', ')
  puts result.to_json
  exit 0
rescue StandardError => e
  result['status'] = 'failure'
  result['result'] = "ENV: #{ENV.inspect}"
  unlock_notify(result) if defined?(unlock_notify)
  result['_error'] = { msg: e.message, kind: 'service/unlock_puppet', details: {} }
  puts result.to_json
  exit 1
end


files/unlock_puppet.rb
#!/opt/puppetlabs/puppet/bin/ruby

# unlock_puppet/tasks/init.rb and unlock_puppet/files/unlock_puppet.rb are similar,
# except that the task does not interact with the pxp-agent service, and raises errors,
# and the script does not read JSON from STDIN.

require 'json'
require 'facter'
require 'puppet'

####

def read_params
  options = {}
  script = true
  if script == true
    require 'optparse'
    options['force_process'] = 'false'
    options['force_service'] = 'false'
    parser = OptionParser.new do |opts|
      opts.banner = 'Usage: unlock_puppet.rb [options]'
      opts.separator ''
      opts.separator 'Summary: Unlock puppet agent runs exceeding runinterval or runtimeout'
      opts.separator ''
      opts.separator 'Options:'
      opts.separator ''
      opts.on('--force_process', 'Force a kill of the puppet agent process, and delete its lock file') do
        options['force_process'] = 'true'
      end
      opts.on('--force_service', 'Force a restart of the puppet and pxp-agent service') do
        options['force_service'] = 'true'
      end
      opts.on('-h', '--help', 'Display help') do
        puts opts
        puts
        exit 0
      end
    end
    parser.parse!
  end
  options
end

params = read_params

force_process = params['force_process'] == 'true'
force_service = params['force_service'] == 'true'

Puppet.initialize_settings

lockfile      = Puppet[:agent_catalog_run_lockfile]
lastrunreport = Puppet[:lastrunreport]
runinterval   = Puppet[:runinterval] || 1800
runtimeout    = Puppet[:runtimeout]  || 1800

result = {}
report = []

####

def puppet_binary
  (Facter.value(:os)['family'] == 'windows') ? 'puppet' : '/opt/puppetlabs/bin/puppet'
end

def write_log(message)
  message.tr!('"', "'")
  if Facter.value(:os)['family'] == 'windows'
    `powershell.exe " If ([System.Diagnostics.EventLog]::SourceExists(\"Unlock Puppet\") -eq $False) { New-EventLog -Source \"Unlock Puppet\" -LogName Application } "`
    `powershell.exe " Write-EventLog -Source \"Unlock Puppet\" -LogName Application -EntryType Information -EventID 1234 -Message \"unlock_puppet: #{message}\" "`
  else
    `logger -p local3.info "#{message}"`
  end
end

def kill_process(pid)
  return if pid.zero?
  if Facter.value(:os)['family'] == 'windows'
    `taskkill /f /pid #{pid}`
  else
    `kill -9 #{pid}`
  end
end

def start_service(service)
  `#{puppet_binary} resource service #{service} ensure=running`
  $?
end

def stop_service(service)
  if Facter.value(:os)['family'] == 'windows'
    result = `SC QUERY #{service} | FIND "STATE"`
    if $?.exitstatus.zero?
      service_state = result.split(' ').last.chomp
      if service_state == 'START_PENDING' || service_state == 'STOP_PENDING'
        result = `SC QUERYEX #{service} | FIND "PID"`
        if $?.exitstatus.zero?
          service_pid = result.split(' ').last.chomp.to_i
          if service_pid > 0
            kill_process(service_pid)
            return $?
          end
        end
      end
    end
    `SC STOP #{service}`
  else
    `#{puppet_binary} resource service #{service} ensure=stopped`
  end
  $?
end

def service_status(service)
  `#{puppet_binary} resource service #{service}`
end

def service_enabled(status)
  status.downcase.include?('true')
end

def service_running(status)
  status.downcase.include?('running')
end

####

begin
  report << 'checking puppet agent process, and puppet and pxp-agent services'

  if force_service
    report << 'stopping puppet service'
    stop_service('puppet')
    report << 'stopping pxp-agent service'
    stop_service('pxp-agent')
  end

  if File.file?(lockfile)
    run_pid = File.read(lockfile).to_i
    if force_process
      report << 'killing puppet agent process'
      kill_process(run_pid)
      if File.file?(lockfile)
        report << 'deleting puppet agent process lock file'
        File.delete(lockfile) if File.file?(lockfile)
        report << 'unable to delete puppet agent process lock file' if File.file?(lockfile)
      end
    else
      lockfile_age = (Time.now - File.stat(lockfile).mtime).to_i
      if lockfile_age > [runinterval, runtimeout].max
        report << "puppet agent process lock file age #{lockfile_age} exceeds runinterval #{runinterval} or runtimeout #{runtimeout}"
        report << 'killing puppet agent process'
        kill_process(run_pid)
        if File.file?(lockfile)
          report << 'deleting puppet agent process lock file'
          File.delete(lockfile) if File.file?(lockfile)
          report << 'unable to delete puppet agent process lock file' if File.file?(lockfile)
        end
      end
    end
  end

  puppet_service_status = service_status('puppet')
  pxp_agent_service_status = service_status('pxp-agent')

  if force_service
    report << 'starting puppet service'
    command = start_service('puppet')
    report << 'unable to start puppet service' unless command.exitstatus.zero?
    report << 'starting pxp-agent service'
    command = start_service('pxp-agent')
    report << 'unable to start pxp-agent service' unless command.exitstatus.zero?
  else
    # Puppet Service
    if service_enabled(puppet_service_status)
      if service_running(puppet_service_status)
        if File.file?(lastrunreport)
          lastrunreport_age = (Time.now - File.stat(lastrunreport).mtime).to_i
          # Either [runinterval, runtimeout].max or runinterval
          if lastrunreport_age > [runinterval, runtimeout].max
            report << "last run report age #{lastrunreport_age} exceeds runinterval #{runinterval} or runtimeout #{runtimeout}"
            report << 'stopping puppet service'
            stop_service('puppet')
            report << 'starting puppet service'
            command = start_service('puppet')
            report << 'unable to start puppet service' unless command.exitstatus.zero?
          end
        else
          report << 'last run report absent'
          report << 'stopping puppet service'
          stop_service('puppet')
          report << 'starting puppet service'
          command = start_service('puppet')
          report << 'unable to start puppet service' unless command.exitstatus.zero?
        end
      else
        # status = puppet_service_status.match(/ensure.*?,/m).to_s.split("'")[1]
        # report << "puppet service status: #{status}"
        report << 'puppet service not running'
        report << 'stopping puppet service'
        stop_service('puppet')
        report << 'starting puppet service'
        command = start_service('puppet')
        report << 'unable to start puppet service' unless command.exitstatus.zero?
      end
    end
    # PXP Agent Service
    if service_enabled(pxp_agent_service_status)
      unless service_running(pxp_agent_service_status)
        # status = pxp_agent_service_status.match(/ensure.*?,/m).to_s.split("'")[1]
        # report << "pxp-agent service status: #{status}"
        report << 'pxp-agent service not running'
        report << 'stopping pxp-agent service'
        stop_service('pxp-agent')
        report << 'starting pxp-agent service'
        command = start_service('pxp-agent')
        report << 'unable to start pxp-agent service' unless command.exitstatus.zero?
      end
    end
  end

  # To enable notifications, create unlock_notify.rb in this directory, containing a unlock_notify(result) method.
  # The unlock_notify(result) method must accept one parameter: the 'result' hash.
  notify_library = File.join(__dir__, 'unlock_notify.rb')
  if File.file?(notify_library)
    require notify_library
  end

  result['status'] = 'success'
  unlock_notify(result) if defined?(unlock_notify)
  result['result'] = report.join(', ')
  puts result.to_json
  exit 0
rescue StandardError => e
  result['status'] = 'failure'
  result['result'] = "ENV: #{ENV.inspect}"
  unlock_notify(result) if defined?(unlock_notify)
  result['_error'] = { msg: e.message, kind: 'service/unlock_puppet', details: {} }
  puts result.to_json
  exit 1
end
